engine/src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -402,7 +411,21 @@ class WorkflowActor(workflowToStart: WorkflowToStart,
       goto(WorkflowFailedState) using data.copy(lastStateReached = StateCheckpoint(FinalizingWorkflowState, Option(failures)))
     case Event(AbortWorkflowCommand, _) => stay()
   }
-  
+
+  when(MetadataIntegrityValidationState) {
+    case Event(_: MetadataWriteSuccess, _) =>
+      context.parent ! WorkflowActorWorkComplete(workflowId, self, stateData.lastStateReached.state.workflowState)
+      context stop self
+      stay()
+    case Event(MetadataWriteFailure(reason, msgs), _) =>
+      // If we shut down now the WMA will erase this workflow from the workflow store, so try again and hope for
+      // better luck. If we continue to be unable to write the completion message to the DB it's better to leave the
+      // workflow in its current state in the DB than to let the WMA delete it
+      workflowLogger.error(reason, "Unable to complete workflow due to inability to write concluding metadata status. Retrying...")
+      PutMetadataActionAndRespond(msgs, self)
+      stay()
```

Comment:
Is this an infinite retry? I don't see any increment/decrement logic that would bound it.

Reply:
It is. I don't know how better to handle it but I'm open to ideas.

Eg, what happens when Cromwell has failed to write the concluding status to the DB 10 times in a row? Should we give up and delete the workflow anyway? Keep trying? Shut down (on the assumption that failure to write basic metadata to the database is more or less a fatal condition?)

Reply:
Would it be possible that if the WorkflowActor hasn't been able to write the final status after 'X' times, the workflow actor shuts down, which leads to a stale heartbeat, and as a result some other Cromwell runner is able to pick up the workflow and run it? (and I am assuming that since majority of the work for the workflow would have already happened, the new Cromwell runner will probably only have to write the final metadata status)